relay-server/src/actors/upstream.rs

```diff
@@ -133,6 +157,47 @@ impl UpstreamRateLimits {
     }
 }
 
+/// Requests are queued and send to the HTTP connections according to their priorities
+/// High priority messages are sent first and then, when no high priority message is pending,
+/// low priority messages are sent. Within the same priority messages are sent FIFO.
+pub enum RequestPriority {
+    /// High priority, low volume messages (e.g. ProjectConfig, ProjectStates, Registration messages)
+    High,
+    /// Low priority, high volume messages (e.g. Events and Outcomes)
+    Low,
+}
+
+/// An type implementing this trait can be placed in a queue according to its priority.
+/// This trait is implemented by message types that need to be sent via http
+pub trait WithRequestPriority {
+    fn priority() -> RequestPriority;
+}
+
+/// UpstreamRequest objects are queued inside the Upstream actor.
+/// The objects are transformed int HTTP requests, and send to upstream as HTTP connections
+/// become available.
+struct UpstreamRequest {
+    response_sender: oneshot::Sender<Result<ClientResponse, UpstreamRequestError>>,
+    method: Method,
+    path: String,
+    build: Box<dyn Send + FnOnce(&mut ClientRequestBuilder) -> Result<ClientRequest, ActixError>>,
+}
+
+pub struct UpstreamRelay {
+    // receiver queue
+    pump_http_queue_notifications: Option<mpsc::Receiver<TrackedFutureFinished>>,
+    http_request_finished_notifier: mpsc::Sender<TrackedFutureFinished>,
+    backoff: RetryBackoff,
+    config: Arc<Config>,
+    auth_state: AuthState,
+    max_inflight_requests: usize,
+    num_inflight_requests: usize,
+    // high priority messages
+    hp_messages: VecDeque<UpstreamRequest>,
```

Comment:
_nit_: Rename to `high_prio_requests` for clarity. Same below for low prio.

Reply:
ok